### PR TITLE
Picross

### PR DESCRIPTION
**Pull request by hunterdyar**
First working push of Picross, more or less fully featured.

Known bugs are on level select screen, when you select out of bounds. They're simply caused by not having made enough levels yet.

not confident that the code or memory usage is optimized, would appreciate a review at some point. I've only done a handful of projects without a garbage collector.

### CODE REVIEW

File: main/modes/picross/mode_picross.c [hunk 1]
```diff
@@ -0,0 +1,1167 @@
...
+    }
+    
+}
+
+//lol i dont know how to do constructors, im a unity developer! anyway this constructs a source.
+//we use the same data type for rows and columns, hence the bool flag. index is the position in the row/col, and we pass in the level.
+//I guess that could be a pointer? Somebody more experienced with c, code review this and optimize please.
+picrossHint_t newHintFromPuzzle(uint8_t index, bool isRow, picrossSpaceType_t source[10][10])
+{
+    picrossHint_t t;
```
Comment (AEFeinstein) [inline main/modes/picross/mode_picross.c:224]: I had to do some research on this, and I think it's OK.
https://stackoverflow.com/questions/9653072/return-a-struct-from-a-function-in-c
What I would have done, out of habit, is have this function return `picrossHint_t*`, `calloc()` a hint, and return a pointer to that.

I like the comment for the function, but I'd recommend writing it in the doxygen style to describe arguments. There are plugins to autogen those comment blocks

File: main/modes/picross/mode_picross.c [hunk 2]
```diff
@@ -0,0 +1,1167 @@
...
+            hintbox.x1 = hintbox.x1 - (g * (j));
+            //we have to flip the hints around. 
+            if(h == 0){
+                //dont increase j.
+                
+            }else if(h < 10){
+                
+                //drawBox(d,hintbox,c111,false,1);//for debugging. we will draw nothing when proper.
+
+                char letter[1];
```
Comment (AEFeinstein) [inline main/modes/picross/mode_picross.c:941]: I think this needs to be bigger, one char for `h` and one for a null terminator.

File: main/modes/picross/mode_picross.c [hunk 3]
```diff
@@ -0,0 +1,1167 @@
...
+                sprintf(letter, "%d", h);//this function appears to modify hintbox.x0
+                //as a temporary workaround, we will use x1 and y1 and subtract the drawscale.
+                drawChar(d,c555, font->h, &font->chars[(*letter) - ' '], (getHintShift(h)+hintbox.x1-p->drawScale) >> 1, (hintbox.y1-p->drawScale+vHintShift) >> 1);
+                j++;//the index position, but only for where to draw. shifts the clues to the right.
+            }else{
+                //drawBox(d,hintbox,c111,false,1);//same as above
+                char letter[1];
```
Comment (AEFeinstein) [inline main/modes/picross/mode_picross.c:948]: This is definitely gonna write out of bounds if `h>=10`

File: main/modes/picross/mode_picross.c (hunk 2 above)
Comment (AEFeinstein) [reply main/modes/picross/mode_picross.c:941]: If you want to be a little safer, you can always use `snprintf()`

Comment (AEFeinstein) [reply main/modes/picross/mode_picross.c:941]: You could also do `char c = ('0' + h);`, but that only works for one digit numbers

File: main/modes/picross/mode_picross.c [hunk 4]
```diff
@@ -0,0 +1,1167 @@
...
+
+void picrossExitGame(void)
+{
+    //set LED's to off.
+    setLeds(p->offLEDS, NUM_LEDS);
+
+    if (NULL != p)
+    {
+        freeFont(&(p->hint_font));
+        // free(p->puzzle);
```
Comment (AEFeinstein) [inline main/modes/picross/mode_picross.c:1059]: Why no free?

File: main/modes/picross/mode_picross.c [hunk 5]
```diff
@@ -0,0 +1,1167 @@
...
+}
+void saveCompletedOnSelectedLevel(bool completed)
+{
+        //Save Victory
+        int32_t victories = 0;
+        if(p->selectedLevel->index <= 32)//levels 0-31
+        {
+            //Save the fact that we won.
+            
+            readNvs32("picross_Solves0", &victories);
```
Comment (AEFeinstein) [inline main/modes/picross/mode_picross.c:1105]: Keys...

**Comment by AEFeinstein** [comment]
Overall, pretty good for a self-professed Unity dev!

**Comment by AEFeinstein** [comment]
When you're comfortable, I'd like to merge this to `main` and have you work in a branch rather than a fork. It's a lot easier for me to switch between everyone's branches than forks.

File: main/modes/picross/picross_select.c [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    ls->game_font = font;
+    //8 is numLevels
+    ls->levelCount = 8;
+    loadWsg("unknownPuzzle.wsg",&ls->unknownPuzzle);
+    //Load in which levels have been completed.
+    int32_t victories0 = 0;
+    int32_t victories1 = 0;
+    int32_t victories2 = 0;
+    //picross_Solves1 would be for levels 33->64
+    readNvs32("picross_Solves0", &victories0);
```
[the remark this answers is not in the document]
Comment (hunterdyar) [reply main/modes/picross/picross_select.c:50]: I think whatever makes the most sense with the rest of the system. We aren't even sure we will get 32+ levels designed and finished right now, but the goal is 50, so even just get/set for i64 will probably be usable for the short term/pragmatic.

For getting values of arbitrary length, if str would be more useful for others than blob, that seems to be a similar solution. My level-complete-save data is just a 1 or 0 bit at some index position.

File: main/modes/picross/picross_select.c [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    readNvs32("picross_Solves1", &victories1);
+    readNvs32("picross_Solves2", &victories2);
+
+    int32_t v;
+    int j;//bit position of i in appropriate register
+    for(int i = 0;i<ls->levelCount;i++)
+    {
+        //todo: I bet there is a smarter way to do this.
```
[the remark this answers is not in the document]
Comment (hunterdyar) [reply main/modes/picross/picross_select.c:58]: "Not splitting one piece of data into three variables" - me teaching about structs and classes to my students. Oh how the turn tables.

File: main/modes/picross/mode_picross.c (hunk 1 above)
Comment (hunterdyar) [reply main/modes/picross/mode_picross.c:224]: I'm going to leave it as-is for now. Doxygen comments updated. Thanks for the input!

**Comment by AEFeinstein** [comment]
OK, I'm gonna merge this now and make a branch for you here, rather than using your fork. Welcome aboard collaborator!

For the future, my condensed notes are:
* NVS keys picross uses should be `const char[]` somewhere consistent, not typed out each time
* Being restricted to 32 bit NVS read/write isn't great, and when the system allows something better, you may want to take advantage of that. Or not.
* Mode-agnostic drawing functions should be moved to `display.c`. All that stuff's getting optimized too.
* Do some memory analysis at some point. Leaking is bad, but the swadge reboots into each mode, so at least your leak won't hinder someone else's mode.
* Be careful not to write out of bounds (`sprintf` into one byte arrays)

**Comment by AEFeinstein** [comment]
Here's your branch, go nuts!
https://github.com/AEFeinstein/Super-2023-Swadge-FW/tree/picross

**Pull request merged**